### ISSUE TITLE
multiple corpuses + tracing for extractive vector generation

### ISSUE THREAD

**Issue by raulk** [filecoin-project/test-vectors#150]
**_Capturing discussion from colo._**

We are maturing our tooling for extracting message-class test vectors from live networks.

This is not an easy task, and the main complexity here is hitting the optimal tradeoff between vector size (precondition state data) and future-proofing.

## Current state (coarse-grained vectors; capturing the full state of accessed actors)

Right now, our extraction tool traces the actors that are accessed during the execution of a message, and masks/trims the state tree to only include the effectively-accessed actors (and their IPLD links) in the precondition state tree.

This produces highly variable vector sizes, depending on the logic of the actual actor method that is executed. Some vectors are 800kb large, others are 11mb large.

The heaviest actors are the Storage Power and the Market actors. If during the execution of a message, those actors are accessed, the vector is almost guaranteed to have a large footprint, due to all the claims, deals, etc. data that is being saved.

## Alternative approach (fine-grained vectors; capturing only effectively-accessed CIDs)

An alternative approach consists of tracing the actual CIDs that are accessed during the execution of a message, and saving ONLY those in the precondition state tree CAR.

We'd need to use a special CAR reader that ignores links for missing CIDs, and CAR writer that filters the writes through a CID allowlist (captured by the tracer).

That way we'll end up with broken links inside the CAR, but it won't matter because those links are never accessed, at least not in the actors version we're testing. Lo and behold, here comes the future-proofing conundrum!

## Future-proofing notes

Actor logic may change in the future in ways that are non consensus breaking (therefore don't justify the scrapping of a test vector), yet it starts accessing state that was not captured in the vector (e.g. map entries inside the state of an actor that already participates in the message). That's when the broken links in fine-grained vectors will start hurting us, and would render the fine-grained vector useless.

OTOH, if the actor logic changes in a way that processing a message now performs more internal sends and therefore requires actor state that we never captured, that is _anyway_ very likely to lead to a consensus-breaking change, because the gas costs will change, even if state modifications are unaffected (i.e. the postcondition state tree still stands). Thus a new test vector is required anyway.

## Striking balance

I consider fine-grained vectors to be a highly-desired optimisation that makes the corpus lightweight. We can't afford to have GiB-large corpuses!

On the other hand, coarse-grained vectors are the safer say to go, to insulate against the kinds of actor changes that are feasible without consensus breakage.

In other words, coarse-grained vectors have a much longer lifetime than finer-grained vectors, **but in the majority of cases, finer-grained vectors can be regenerated from coarser grained vectors. I think this is a property we can leverage to strike the optimum solution.**

## Solution proposal

Drawing from the above, it appears that a desirable solution would be managing a two-tier corpus:

* Tier 1: lightweight corpus
	* light checkout / smaller footprint; test vectors in the order of 1-3 digit KiB.
    * contains all vectors generated through the generative approach (builders) + fine-grained test vectors.
* Tier 2: heavy corpus
    * larger footprint; test vectors in the order of 1-2 digit MiB.
    * only used as a fallback for healing the Tier 1 corpus on logic changes, without having to resort to fully regenerating the vector from chain (which would require an archive node -- not always available/practical)
    * contains all vectors generated through the generative approach (builders) + coarse-grained test vectors.

This would require:

- [ ] extraction tooling that supports various masking modes (actor-level, CID-level).
- [ ] tooling to regenerate a fine-grained vector from a coarse-grained vector, using a concrete version of lotus/specs-actors.
- [ ] CI jobs to know when fine-grained vectors are rendered useless, on a particular commit.
- [ ] record tracing information on fine-grained vectors so that we know what coarse-grained vector can regenerate it. Also record message CID + network name.

**Comment by austinabell** [comment]
I believe that the lightweight tests are fine, because if our client ever did request an extra node/cid from the blockstore, there would be a gas discrepancy anyway. This might make debugging harder in some cases, but I believe in others it will make it much easier (because tracking gas usage diff is not so simple).

This approach seems reasonable to me. Yes, lightweight corpus might more easily break older versions, but if there was any difference it would break the result of our runner anyway (unless some retrieval outside of actor execution failed). Intuitively, I don't really see the concern with the lightweight vectors.

As for extracting which cids are used, should be pretty simple with just wrapping a blockstore and keeping a cache of read/ write cids (although might be a little nuanced to ensure that only blocks that are added and didn't exist before are excluded, maybe could check against the pre state CAR).

**Comment by anorth** [comment]
I am a bit skeptical of the value of extracted vectors anyway. There's a lot of value in a _very_ coarse vector of "revalidate the entire chain history", but after that I think the balance of benefit tips away quickly.

Anything that could be extracted as a light vector would I think be better exercised by replicating the behaviour with an explicitly constructed message/tipset/etc vector. It's a lot of work to get really broad coverage that way, but for blanket coverage we should just revalidate whole chain segments anyway, with the large pre-state.

I don't think the large state necessarily is a barrier. We don't need to check that state in to git, just a hash and a reference to where we have archived that state for fetching during the (infrequent, not in dev tight loop) running of those expensive tests.

So in summary:
- I prefer programatic tests to any narrow vector we would explicitly extract
- prefer coarse test for blanket coverage

**Comment by alanshaw** [comment]
> "revalidate the entire chain history"

...on that idea -> **Continuous Conformance**

Ok, I haven't given this a _whole_ lot of thought _but_...

If implementations can provide a driver, can we set up a service whereby vectors are automatically extracted from a live network in realtime, and run them against the implementation? Any conformance failure would stop the process until a push to master (which would redeploy the implementation to the service) and testing would continue, catching up to the latest head.

It has the potential to capture regressions, validate concensus breakages, capture concensus transitions and allows implementations to verify they're conformant...continuously!

**Comment by raulk** [comment]
This has been delivered, tvx has been migrated to Lotus, and the extracted corpus now lives under `corpus/extracted` in this repo.
